Rebased onto current `main` (now on top of #281 `0de8264`) — clean, no conflicts (#281 touched `ingest_protection.py`, disjoint from `engine.py`). Force-pushed `6b77c7e`.

Focused validation rerun on the rebased branch:

- **Regression test** `TestEngineCompress::test_assemble_context_summary_role_is_user_after_system_anchor` — RED on the rebased base (reverting just `engine.py` to the parent → `AssertionError: - user / + assistant`), GREEN with the fix.
- `pytest tests/test_lcm_core.py tests/test_lcm_engine.py tests/test_packaging_install.py -q` → **672 passed, 1 skipped**, plus one flaky timing test (`TestSessionRollover::test_on_session_end_returns_quickly_under_real_sqlite_writer_lock`) that passes in isolation and on stock `main` — it's load-dependent under the full parallel run, unrelated to this change.
- `python -m compileall -q .` ✓
- `python -m py_compile scripts/import_lossless_claw.py` ✓
- `bash -n scripts/install.sh scripts/update.sh` ✓
- `git diff --check` ✓

Ready for review on the current branch.

--- a/engine.py
+++ b/engine.py
@@ -4285,7 +4285,15 @@ class LCMEngine(ContextEngine):
         # Collect DAG summaries — highest depth first for context hierarchy
         summary_parts: list[str] = []
         last_role = result[-1].get("role", "system") if result else "system"
-        summary_role = "assistant" if last_role != "assistant" else "user"
+        if result and result[-1].get("role") == "system":
+            # The system prompt is the only leading anchor, so the summary
+            # becomes the first message reaching the provider. Anthropic
+            # extracts the system prompt into a separate field and requires
+            # messages[0] to be role "user"; an assistant summary here is
+            # rejected with HTTP 400 after the second compaction.
+            summary_role = "user"
+        else:
+            summary_role = "assistant" if last_role != "assistant" else "user"
         if anchor_part is not None:
             anchor_msg = {"role": summary_role, "content": anchor_part}
             if summary_budget is None or count_message_tokens(anchor_msg) <= summary_budget:

--- a/engine.py
+++ b/engine.py
@@ -4285,12 +4285,13 @@ class LCMEngine(ContextEngine):
         # Collect DAG summaries — highest depth first for context hierarchy
         summary_parts: list[str] = []
         last_role = result[-1].get("role", "system") if result else "system"
-        if result and result[-1].get("role") == "system":
-            # The system prompt is the only leading anchor, so the summary
-            # becomes the first message reaching the provider. Anthropic
-            # extracts the system prompt into a separate field and requires
-            # messages[0] to be role "user"; an assistant summary here is
-            # rejected with HTTP 400 after the second compaction.
+        if not result or result[-1].get("role") == "system":
+            # The summary becomes the first provider-visible message: either no
+            # leading anchor exists (gateway-style assembly) or the system
+            # prompt is the only anchor, which Anthropic extracts into a
+            # separate field. Either way messages[0] must be role "user"; an
+            # assistant summary here is rejected with HTTP 400 after the second
+            # compaction.
             summary_role = "user"
         else:
             summary_role = "assistant" if last_role != "assistant" else "user"

--- a/tests/test_lcm_core.py
+++ b/tests/test_lcm_core.py
@@ -3737,8 +3737,14 @@ class TestAssemblyBudgetSelection:
         assert "KEEP_USER_DECISION" in contents
         assert "Latest compact status" in contents
         assert "oversized assistant tool chatter" not in contents
+        # The preserved objective is replayed as scaffolding (identified by its
+        # content prefix, not its role), so it must never appear as a raw
+        # user-role turn that could be re-ingested as a durable row.
         assert not any(
-            msg.get("role") == "user" and "KEEP_USER_DECISION" in str(msg.get("content", ""))
+            msg.get("role") == "user"
+            and "[Current user objective preserved from compacted history]"
+            not in str(msg.get("content", ""))
+            and "KEEP_USER_DECISION" in str(msg.get("content", ""))
             for msg in assembled
         )
 
@@ -3802,8 +3808,14 @@ class TestAssemblyBudgetSelection:
             and "KEEP_USER_DECISION" in str(msg.get("content", ""))
             for msg in assembled
         )
+        # The preserved objective is replayed as scaffolding (identified by its
+        # content prefix, not its role), so it must never appear as a raw
+        # user-role turn that could be re-ingested as a durable row.
         assert not any(
-            msg.get("role") == "user" and "KEEP_USER_DECISION" in str(msg.get("content", ""))
+            msg.get("role") == "user"
+            and "[Current user objective preserved from compacted history]"
+            not in str(msg.get("content", ""))
+            and "KEEP_USER_DECISION" in str(msg.get("content", ""))
             for msg in assembled
         )
 

--- a/tests/test_lcm_engine.py
+++ b/tests/test_lcm_engine.py
@@ -4076,7 +4076,7 @@ class TestEngineCompress:
             "[Expand for details: backed prior]"
         )
         messages = [
-            {"role": "assistant", "content": active_summary_marker},
+            {"role": "user", "content": active_summary_marker},
             {"role": "user", "content": "fresh tail question"},
             {"role": "assistant", "content": "fresh tail answer"},
         ]
@@ -4164,6 +4164,41 @@ class TestEngineCompress:
         ]
 
         result = engine._assemble_context(system_msg, tail)
+
+        non_system = [m for m in result if m.get("role") != "system"]
+        assert non_system, "expected at least one non-system message"
+        assert non_system[0]["role"] == "user"
+        summary_block = next(
+            m for m in result
+            if isinstance(m.get("content"), str) and "prior work summary" in m["content"]
+        )
+        assert summary_block["role"] == "user"
+
+    def test_assemble_context_summary_role_is_user_without_system_anchor(self, engine):
+        """Gateway-style assembly (no system anchor) must not lead with ``assistant``.
+
+        ``_assemble_context`` can be called without a system message, e.g. a
+        gateway session that starts directly with user turns. With no leading
+        anchor the DAG summary becomes ``messages[0]`` itself, so it must be
+        ``role="user"`` for the same reason Anthropic rejects a leading
+        ``assistant`` entry with HTTP 400.
+        """
+        engine._dag.add_node(SummaryNode(
+            session_id=engine._session_id,
+            depth=0,
+            summary="prior work summary",
+            token_count=50,
+            source_token_count=5000,
+            source_ids=[],
+            source_type="messages",
+            created_at=1.0,
+        ))
+        tail = [
+            {"role": "user", "content": "do the thing"},
+            {"role": "assistant", "content": "working on it"},
+        ]
+
+        result = engine._assemble_context(None, tail)
 
         non_system = [m for m in result if m.get("role") != "system"]
         assert non_system, "expected at least one non-system message"

--- a/tests/test_lcm_engine.py
+++ b/tests/test_lcm_engine.py
@@ -4137,6 +4137,43 @@ class TestEngineCompress:
         assert result[0]["content"][-1]["type"] == "text"
         assert "Lossless Context Management" in result[0]["content"][-1]["text"]
 
+    def test_assemble_context_summary_role_is_user_after_system_anchor(self, engine):
+        """The summary must not be the leading non-system message as ``assistant``.
+
+        After compaction the system prompt is the only leading anchor, so the
+        DAG summary becomes the first message reaching the provider. Anthropic
+        extracts the system prompt into a separate field and rejects a request
+        whose first ``messages`` entry is ``assistant`` with HTTP 400 (the
+        misleading ``tool_use ids ... without tool_result`` error). The summary
+        block must therefore be assigned ``role="user"``.
+        """
+        engine._dag.add_node(SummaryNode(
+            session_id=engine._session_id,
+            depth=0,
+            summary="prior work summary",
+            token_count=50,
+            source_token_count=5000,
+            source_ids=[],
+            source_type="messages",
+            created_at=1.0,
+        ))
+        system_msg = {"role": "system", "content": "You are an agent."}
+        tail = [
+            {"role": "user", "content": "do the thing"},
+            {"role": "assistant", "content": "working on it"},
+        ]
+
+        result = engine._assemble_context(system_msg, tail)
+
+        non_system = [m for m in result if m.get("role") != "system"]
+        assert non_system, "expected at least one non-system message"
+        assert non_system[0]["role"] == "user"
+        summary_block = next(
+            m for m in result
+            if isinstance(m.get("content"), str) and "prior work summary" in m["content"]
+        )
+        assert summary_block["role"] == "user"
+
     def test_compress_preserves_latest_user_request_outside_fresh_tail(self, tmp_path, monkeypatch):
         """The latest real user request anchors the task even after tool-heavy turns.
 
@@ -13897,7 +13934,11 @@ class TestEngineTools:
         compressed = engine.compress(messages)
 
         assert compressed[0]["role"] == "system"
-        assert compressed[1]["role"] == "assistant"
+        # The summary follows the system prompt, so it must be role "user" —
+        # an assistant summary as the first non-system message is rejected by
+        # Anthropic (it extracts the system prompt and requires messages[0]
+        # to be "user").
+        assert compressed[1]["role"] == "user"
         assert "Recent Summary" in compressed[1]["content"]
         stored_tool = next(row for row in engine._store.get_range("test-session") if row["role"] == "tool")
         assert stored_tool["content"].startswith("[GC'd externalized tool output:")

